### PR TITLE
Testing /remote.php/dav/meta/$fileId/v

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Description
Follow up of https://github.com/owncloud/core/pull/29488
This PR will add more integration tests for the meta dav resource

## How Has This Been Tested?
using behat

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tests/integration/features/dav-versions.feature [hunk 1]
```diff
@@ -1,28 +1,55 @@
...
+    And user "user0" uploads file with content "12345" to "/davtest.txt"
+    And the version folder of file "/davtest.txt" for user "user0" contains "1" elements
+    When user "user0" restores version nr "1" of file "/davtest.txt"
+    Then downloading file "davtest.txt"
+    And downloaded content should be "123"
+
+  Scenario: User cannot access meta folder of a file which is owned by somebody else
+    Given user "user1" exists
+    And user "user0" uploads file with content "123" to "/davtest.txt"
+    And we save it into "FILEID"
```
Comment (SergioBertolinSG) [inline tests/integration/features/dav-versions.feature]: What is doing this step?

We have functions to get the id of a file, maybe adding this extension rdx/behat-variables can be avoided.

Here we use a variable to store file id:
https://github.com/owncloud/core/blob/master/tests/integration/features/webdav-related-new-endpoint.feature#L358

To get the file id we can use this function:
https://github.com/owncloud/core/blob/master/tests/integration/features/bootstrap/WebDav.php#L1141

File: tests/integration/features/bootstrap/WebDav.php [hunk 2]
```diff
@@ -1168,4 +1179,18 @@ public function userChecksFileIdForPath($user, $path) {
 		$currentFileID = $this->getFileIdForPath($user, $path);
 		PHPUnit_Framework_Assert::assertEquals($currentFileID, $this->storedFileID);
 	}
+
+	/**
+	 * @When user :user restores version nr :versionIndex of file :path
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/WebDav.php]: "nr" means number? then it should be "number", no need to put "nr".

**Review by SergioBertolinSG — CHANGES_REQUESTED** [review]

File: tests/integration/features/dav-versions.feature (hunk 1 above)
Comment (DeepDiver1975) [reply tests/integration/features/dav-versions.feature]: Ah - didn't know that .... but this is causing another request to the server. Good? bad? ...

File: tests/integration/features/bootstrap/WebDav.php (hunk 2 above)
Comment (DeepDiver1975) [reply tests/integration/features/bootstrap/WebDav.php]: Sounds reasonable - will adjust ... or maybe even code different

File: tests/integration/features/dav-versions.feature (hunk 1 above)
Comment (SergioBertolinSG) [reply tests/integration/features/dav-versions.feature]: I think it is better than adding another dependency.

Comment (DeepDiver1975) [reply tests/integration/features/dav-versions.feature]: But I think this step is incomplete 'And user "user0" stores id of file "/textfile0.txt"'

How to save multiple file ids? How to check multiple fileids in follow up steps?

Comment (DeepDiver1975) [reply tests/integration/features/dav-versions.feature]: Furthermore I explicitly need to reference the fileid in the url in a later step

```
  Scenario: User cannot access meta folder of a file which is owned by somebody else
    Given user "user1" exists
    And user "user0" uploads file with content "123" to "/davtest.txt"
    And we save it into "FILEID"
    And as an "user1"
    When sending "PROPFIND" with exact url to "/remote.php/dav/meta/<<FILEID>>"
    Then the HTTP status code should be "404"

Comment (SergioBertolinSG) [reply tests/integration/features/dav-versions.feature]: > How to save multiple file ids? How to check multiple fileids in follow up steps?

Currently only works with one.

About the following steps, you could add a new step like

When getting metadata of last file id stored

And keep the last part about the http status.

**Comment by DeepDiver1975** [comment]
> When getting metadata of last file id stored
> 
> And keep the last part about the http status.

but then I have to code a hell of new steps vs. re-using an existing step and get simple string interpolation.

I'm not convinced ...

**Comment by SergioBertolinSG** [comment]
OK, add the dependency then, maybe we can use it in the future.

The problem could be if we upgrade behat version at some point and somehow this extension doesn't work with the newer version, making us fixing it upstream or having to remove it.

File: tests/integration/features/bootstrap/WebDav.php [hunk 3]
```diff
@@ -1168,4 +1179,18 @@ public function userChecksFileIdForPath($user, $path) {
...
 		PHPUnit_Framework_Assert::assertEquals($currentFileID, $this->storedFileID);
 	}
+
+	/**
+	 * @When user :user restores version index :versionIndex of file :path
+	 * @param string $user
+	 * @param int $versionIndex
+	 * @param string $path
+	 */
+	public function userRestoresVersionNrOfFile($user, $versionIndex, $path) {
```
Comment (SergioBertolinSG) [inline tests/integration/features/bootstrap/WebDav.php]: The function name has the old name.

**Review by SergioBertolinSG — CHANGES_REQUESTED** [review]

File: tests/integration/features/bootstrap/WebDav.php (hunk 3 above)
Comment (DeepDiver1975) [reply tests/integration/features/bootstrap/WebDav.php]: k - will adjust .... but it actually has no meaning - right?

Comment (DeepDiver1975) [reply tests/integration/features/bootstrap/WebDav.php]: I mean the function is detected by the regex in the phpdoc - unlike ruby/cucumber where the function name is used .... anyway

Comment (SergioBertolinSG) [reply tests/integration/features/bootstrap/WebDav.php]: Yes, just a cosmetic change.

Comment (DeepDiver1975) [reply tests/integration/features/bootstrap/WebDav.php]: done

**Review by SergioBertolinSG — APPROVED** [review]

**Comment by phil-davis** [comment]
@DeepDiver1975 @SergioBertolinSG should these tests be backported?

I noticed because I am looking at ``composer.json`` diffs between ``master`` and ``stable10``.

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 @SergioBertolinSG should these tests be backported?
> 
> I noticed because I am looking at composer.json diffs between master and stable10.

yes please

**Comment by phil-davis** [comment]
Assigned myself so I find this again. I will backport after the latest composer PR in stable10 gets merged (no point backporting now when I know there will be more conflict to come)

**Comment by phil-davis** [comment]
See discussion in #30429 - backport here is waiting on/dependent on if other related meta dav stuff is backported. Leaving backport label here for now, so it gets noticed every now and then.

**Comment by phil-davis** [comment]
This code is all backported in #31050 which has been merged. Removing the backport-request label.

**Pull request merged**